sips/sip-023/sip-023-emergency-fix-traits.md

```diff
@@ -0,0 +1,115 @@
+# Preamble
+
+SIP Number: 023
+
+Title: Emergency Fix to Trait Invocation Behavior
+
+Authors:
+    Aaron Blankstein <aaron@hiro.so>,
+    Brice Dobry <brice@hiro.so>,
+    Jude Nelson <jude@stacks.org>,
+
+Consideration: Technical, Governance
+
+Type: Consensus
+
+Status: Draft
+
+Created: 1 May 2023
+
+License: BSD 2-Clause
+
+Sign-off: 
+
+Discussions-To: https://github.com/stacksgov/sips
+
+# Abstract
+
+On 1 May 2023, it was discovered that pre-2.1 contracts exposing public methods with
+trait arguments could not be invoked with previously working trait implementating
+contract arguments.
+
+This bug was caused by the activation of Stacks Epoch 2.2.
+
+This SIP proposes an **immediate consensus-breaking change** to
+introduce a new Stacks Epoch 2.3 that corrects this regression.
+
+**This SIP proposes a Bitcoin activation height of 788,287**
+
+# Epoch 2.2 Bug Behavior
+
+Stacks Epoch 2.1 introduced a new type checker and type system which
+impacts trait invocations. In order for existing contracts to remain
+compatible, their types must be _canonicalized_. The canonicalization
+method performed an exact check for the current epoch:
+
+```rust
+    pub fn canonicalize(&self, epoch: &StacksEpochId) -> TypeSignature {
+        match epoch {
+            StacksEpochId::Epoch21 => self.canonicalize_v2_1(),
+            _ => self.clone(),
+        }
+    }
+```
+
+Therefore, a Pre-2.1 with trait arguments that is invoked in Epoch 2.2
+will fail to canonicalize its trait arguments, and abort with a
+runtime error.
```

Comment:
It would be more accurate to say that a contract call transaction will be rejected by the mempool with a type error, and a read-only call will fail with a runtime error.

Reply:
With the new 2.1 "all errors are runtime errors", a miner could choose to include the transaction and it would have the runtime error (just like when wrapping), so I think it makes sense both ways.

Reply:
updated the change made to address @obycode comment with the remark about a miner including the tx